### PR TITLE
add some more documentation for e2e-tests

### PR DESCRIPTION
**Pull request by rajatjindal**
@michelleN,  as discussed I've added some documentation about the key concepts for types/functions used in e2e-tests. Please let me know if adding any more info will be useful.

### CODE REVIEW

File: tests/README.md [hunk 1]
```diff
@@ -42,11 +42,63 @@ docker compose                                          \
...
-`crates/e2e-testing`     - All the test framework/utilities that are required for `e2e-tests`
-`tests/testcases/mod.rs` - All the testcase definitions should be added here.
-`tests/spinup_tests.rs`  - All tests that we want to run with `Spin Up` should be added here
-`tests/testcases/<dirs>` - The testcases which require corresponding `spin app` pre-created should be added here
+`crates/e2e-testing`     - All the test framework/utilities that are required for `e2e-tests`<br/>
+`tests/testcases/mod.rs` - All the testcase definitions should be added here.<br/>
+`tests/spinup_tests.rs`  - All tests that we want to run with `Spin Up` should be added here<br/>
+`tests/testcases/<dirs>` - The testcases which require corresponding `spin app` pre-created should be added here<br/>
 
+## key concepts and types
```
Comment (radu-matei) [inline tests/README.md:50]: nit: capitalize the headers.

File: tests/README.md [hunk 2]
```diff
@@ -42,11 +42,63 @@ docker compose                                          \
...
+
+### [trait Controller](../crates/e2e-testing/src/controller.rs#L12)
+
+This defines a trait which can be implemented by different deployment models (e.g. [`spin up`](../crates/e2e-testing/src/spin_controller.rs#L15) or `Fermyon cloud`). Using this we can reuse same testcases, which can be executed against these different deployment models. e.g. they may chose to have different way to start/stop the spin apps.
```
Comment (radu-matei) [inline tests/README.md:54]: nit: "Fermyon *C*loud"

Comment (rajatjindal) [reply tests/README.md]: thnx Radu. done

File: tests/README.md (hunk 1 above)
Comment (rajatjindal) [reply tests/README.md]: thnx, done

File: tests/README.md [hunk 3]
```diff
@@ -42,11 +42,63 @@ docker compose                                          \
...
+`crates/e2e-testing`     - All the test framework/utilities that are required for `e2e-tests`<br/>
+`tests/testcases/mod.rs` - All the testcase definitions should be added here.<br/>
+`tests/spinup_tests.rs`  - All tests that we want to run with `Spin Up` should be added here<br/>
+`tests/testcases/<dirs>` - The testcases which require corresponding `spin app` pre-created should be added here<br/>
 
+## Key concepts and types
+
+### [trait Controller](../crates/e2e-testing/src/controller.rs#L12)
+
+This defines a trait which can be implemented by different deployment models (e.g. [`spin up`](../crates/e2e-testing/src/spin_controller.rs#L15) or `Fermyon Cloud`). Using this we can reuse same testcases, which can be executed against these different deployment models. e.g. they may chose to have different way to start/stop the spin apps.
```
Comment (vdice) [inline tests/README.md:54]:
```suggestion
This defines a trait which can be implemented by different deployment models (e.g. [`spin up`](../crates/e2e-testing/src/spin_controller.rs#L15) or `Fermyon Cloud`). Using this we can reuse the same testcases, which can be executed against these different deployment models (e.g. they may choose to have different way to start/stop the spin apps.)
```

File: tests/README.md [hunk 4]
```diff
@@ -42,11 +42,63 @@ docker compose                                          \
...
+
+### [TestCase](../crates/e2e-testing/src/testcase.rs#L22)
+
+This help us configure the different scenarios/steps which are required for running a specific test app. e.g. it has [TestCase.trigger_type](../crates/e2e-testing/src/testcase.rs#L42) which tells the type of `trigger` a particular test app uses. and similarly it has [TestCase.plugins](../crates/e2e-testing/src/testcase.rs#L53) which tell the pre-requisite [`plugins`](https://developer.fermyon.com/spin/plugin-authoring) that are required to run this test app.
```
Comment (vdice) [inline tests/README.md:58]:
```suggestion
This helps us configure the different scenarios/steps which are required for running a specific test app. For example, [TestCase.trigger_type](../crates/e2e-testing/src/testcase.rs#L42) indicates the `trigger` a particular test app uses and [TestCase.plugins](../crates/e2e-testing/src/testcase.rs#L53) indicates which prerequisite [`plugins`](https://developer.fermyon.com/spin/plugin-authoring) are required to run this test app.
```

**Review by vdice — APPROVED** [review]
Helpful docs additions!

All tweaks optional, except perhaps the one or two typos.

**Comment by rajatjindal** [comment]
thanks Vaughn, I've committed all the suggestions.

**Comment by rajatjindal** [comment]
Hi @vdice , I made an additional change to ignore builds if change is in `tests/README.md` file.

**Review by vdice — APPROVED** [review]

**Pull request merged**